### ISSUE TITLE
java.lang.OutOfMemoryError: unable to create native thread

### ISSUE THREAD

**Issue by MircoGnuva** [trinodb/trino#15732]
Hi,
We use Trino deployed on Kubernetes cluster with the following specification:
- One coordinator pod
- 5 worker pods

Each node has the following limits:
- 4 CPUs 
- Memory: 4 GB for the coordinator pod and 6 GB for the workers ones 

JVM configuration:
```
  -server

  -Xmx4G

  -XX:+UseG1GC

  -XX:G1HeapRegionSize=32M

  -XX:+UseGCOverheadLimit

  -XX:+ExplicitGCInvokesConcurrent

  -XX:+HeapDumpOnOutOfMemoryError

  -XX:+ExitOnOutOfMemoryError

  -Djdk.attach.allowAttachSelf=true

  -Djavax.net.ssl.trustStore=/etc/trino/certs/cacerts

  -Djavax.net.ssl.trustStorePassword=redacted
```

We have updated from version 388 to 403 and now the following exception is raised when a query (that on works 388) is executed:
java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached.

We tried to increase the resources:
- CPUs up to 8 for each pod
- Memory up to 12 GB for each pod

We also tuned the JVM Xmx parameter up to 12GB.

Any suggestion?

**Comment by amir-upstream** [comment]
we are experiencing  a very similar issue, upgraded from 393 to 403, and started getting :
java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached.

Our queries / data hasn't really changed. we are running the same workloads.

running a 1 coordinator and 5 workers (6 cpu each) with:
    maxHeapSize: "48G"
    memoryHeadroomPerNode: "16GB"

**Comment by daha** [comment]
I had the same issue on a Openshift 4.10 cluster. When I updated from 397 to 406 or 418. 

I did some investigation and I found that I got the error when I reached 1024 threads [1]. I found that it was related to some cgroup limit [2] which prevents pods to have more than 1024 threads. 
```sh
$ cat /sys/fs/cgroup/pids/pids.max 
1024
```
When I managed to increase it in my cluster [3] I got rid of the "java.lang.OutOfMemoryError: unable to create native thread" errors. I could see that the thread count could go up to 1300-1400 on my coordinator, after the fix. 

[1]
https://www.mastertheboss.com/jbossas/monitoring/how-to-solve-javalangoutofmemoryerror-unable-to-create-new-native-thread/

[2]
https://access.redhat.com/discussions/4713291

[3]
https://computingforgeeks.com/change-pids-limit-value-in-openshift/
